service/src/main/java/org/broadinstitute/listener/relay/inspectors/SamResourceClient.java

```diff
@@ -0,0 +1,61 @@
+package org.broadinstitute.listener.relay.inspectors;
+
+import java.io.IOException;
+import java.time.Instant;
+import org.broadinstitute.dsde.workbench.client.sam.ApiClient;
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+import org.broadinstitute.dsde.workbench.client.sam.api.ResourcesApi;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cache.annotation.Cacheable;
+
+public class SamResourceClient {
+  private final String samResourceId;
+  private final TokenChecker tokenChecker;
+  private final ApiClient samClient;
+
+  private final String SAM_RESOURCE_TYPE = "controlled-application-private-workspace-resource";
+
+  private final Logger logger = LoggerFactory.getLogger(SamResourceClient.class);
+
+  public SamResourceClient(String samResourceId, ApiClient samClient, TokenChecker tokenChecker) {
+    this.samResourceId = samResourceId;
+    this.tokenChecker = tokenChecker;
+    this.samClient = samClient;
+  }
+
+  // Should only be used in checkCachedPermission, but making it public so that we can test it
+  @Cacheable("expiresAt")
+  public Instant checkWritePermission(String accessToken) {
+    try {
+      var oauthInfo = tokenChecker.getOauthInfo(accessToken);
+      if(oauthInfo.expiresAt().isPresent()) {
+        samClient.setAccessToken(accessToken);
+        var resourceApi = new ResourcesApi(samClient);
+
+        var res = resourceApi.resourcePermissionV2(SAM_RESOURCE_TYPE, samResourceId, "write");
+        if(res)
+          return oauthInfo.expiresAt().get();
```

Comment:
This is totally a code style issue, but I would write this so that all cases fall through to the end of the method except for this one. They all return the same value, so only one `return` is needed. IMO this makes it clearer that all returns are the same.

Reply:
in Scala, I totally would..but how would you do that in java

Reply:
Something like
```
  public Instant checkWritePermission(String accessToken) {
    try {
      var oauthInfo = tokenChecker.getOauthInfo(accessToken);
      if (oauthInfo.expiresAt() != null) {
        samClient.setAccessToken(accessToken);
        var resourceApi = new ResourcesApi(samClient);

        if (resourceApi.resourcePermissionV2(SAM_RESOURCE_TYPE, samResourceId, "write")) {
          return oauthInfo.expiresAt();
        }
      } else {
        logger.error("Token expired " + oauthInfo.error());
      }
    } catch (IOException | InterruptedException e) {
      logger.error("Fail to check token info", e);
    } catch (ApiException e) {
      logger.error("Fail to check Sam permission", e);
    } catch (Exception e) {
      logger.error("Fail for unknown reasons", e);
    }
    return Instant.EPOCH;
  }
```